ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+method, however the length of a signature is unbounded. Rather than
+consume program space with such strings, a method selector is used to
+identify methods in calls. A method selector is the first four bytes of
+the SHA-512/256 hash of the method signature.
+
+For example, the method selector for a method named `add` which takes
+two uint64 parameters and returns a uint128 can be computed as
+follows:
+
+```
+Method signature: add(uint64,uint64)uint128
+SHA-512/256 hash (in hex): 8aa3b61f0f1965c3a1cbfa91d46b24e54c67270184ff89dc114e877b1753254a
+Method selector (in hex): 8aa3b61f
+```
+
+
+#### Method Description
+
+A method description provides further information about a method
+beyond its signature. This description is encoded in JSON and consists
+of a method's name, description, arguments (their types, names, and
+descriptions), and return type (and possible description for the
+return type). From this structure, the method's signature and selector
+can be calculated. The Algorand SDKs will provide convenience
+functions to calculate signatures and selectors from such JSON files.
+
+These details will enable high-level languages and dapps/wallets to
+properly encode arguments and call methods, as well as offer users a
+rich description for a method and its arguments. This description can
+populate UIs in dapps, wallets, and explorers with description of
+parameters, as well as populate information about methods in IDEs for
+HLLs.
+
+The JSON structure for such an object is:
+
+```
+interface Method {
+  name: string,
+  desc?: string,
+  args: Array<{ name: string, type: string, desc?: string }>,
+  returns?: { type: string, desc?: string }
+}
+```
+
+For example:
+
+```
+{
+  "name": "add",
+  "desc": "Calculate the sum of two 64-bit integers",
+  "args": [
+    { "name": "a", "type": "uint64", "desc": "..." },
+    { "name": "b", "type": "uint64", "desc": "..." }
+  ],
+  "returns": { "type": "uint128", "desc": "..." }
+}
+```
+
+
+### Interfaces
+
+An interface is a logically grouped set of methods. All methods in an
+interface must be unique (specifically, they must have unique method
+selectors). Method names in Interfaces **MUST NOT** begin with an
+underscore. A smart contract implements an interface if it supports
+all of the methods from that interface. A smart contract may implement
+zero, one, or multiple interfaces.
+
+#### Interface Description
+
+An interface description is a JSON object containing the JSON
+descriptions for each of the methods in the interface.
+
+The JSON structure for such an object is:
+
+```
+interface Interface {
+  name: string,
+  methods: Array<Method>
+}
+```
+
+For example:
+
+```
+{
+  "name": "Calculator",
+  "methods": [
+    { "name": "add", "args": [...], ... },
+    { "name": "multiply", "args": [...], ... }
+  ]
+}
+```
+
+### Contracts
+
+An contract is the complete set of methods that an app implements. It
+is similar to an interface, but may include further details about the
```

Comment:
We started with Interfaces, and added Contracts as a way to specify some concrete things (_optIn, _closeOut, appId) that don't make sense for an Interface (which is presumably documented in some later ARC - "ARC-6253 : Multi-colored NFTs".

Such interfaces are not allowed to declare how you opt into them, but a particular implementation (Contract) can.

I think we're ok with the potential for clashes (they only occur on name + types).  But optIn/closeOut are special - there the clash would occur because of the same *name*, (not including args). And the meaning would be nonsensical - optin using these two different argument specifications?

I don't hate adding Interface name to the selector, but didn't see enough value to depart from the Ethereum way of naming. I'd like to simply say "This contract has add(uint64,uint64):uint64" and not have to decide if it really belongs in an Interface, which would mean its selector changes.

Reply:
That's all fine and that's why I led with "What's the value of interfaces?" Either they should provide something extra over Ethereum, by being a consistent convention present from the beginning, or they don't really mean anything.